src/main/kotlin/com/geekbeast/rhizome/jobs/HazelcastJobService.kt

```diff
@@ -0,0 +1,231 @@
+import com.hazelcast.map.IMap
+import com.hazelcast.query.Predicate
+import com.hazelcast.query.Predicates
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Service
+import java.util.*
+import java.util.concurrent.Callable
+
+const val JOBS_MAP = "_rhizome_jobs_"
+private const val JOB_STATUS = "status"
+private const val JOBS_EXECUTOR = "_rhizome_job_service_"
+
+/**
+ *
+ * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
+ */
+@Service
+class HazelcastJobService(hazelcastInstance: HazelcastInstance) {
+    companion object {
+        private val logger = LoggerFactory.getLogger(HazelcastJobService::class.java)
+    }
+
+    protected val jobs = hazelcastInstance.getMap<UUID, AbstractDistributedJob<*, *>>(JOBS_MAP)
+    protected val durableExecutor = hazelcastInstance.getDurableExecutorService(JOBS_EXECUTOR)
+    protected val mapper: ObjectMapper = ObjectMappers.getSmileMapper()
+
+
+    @Timed
+    fun resumeJobs(ids: Set<UUID>) {
+        //Rather than pulling the job here, we simply create a wrapper job that
+        ids.forEach { id ->
+            durableExecutor.submitToKeyOwner(object : HazelcastInstanceAware, java.io.Serializable, Callable<Any?> {
+                @Transient
+                private lateinit var jobs: IMap<UUID, AbstractDistributedJob<*, *>>
+                override fun setHazelcastInstance(hazelcastInstance: HazelcastInstance) {
+                    jobs = hazelcastInstance.getMap(JOBS_MAP)
+                }
+
+                override fun call(): Any? {
+                    return jobs[id]?.call()
+                }
+
+            }, id)
+        }
+    }
+
+    @Timed
+    fun <R, S : JobState> submitJob(job: AbstractDistributedJob<R, S>): UUID {
+        require(job.status == JobStatus.PENDING) { "Job status must be pending to submit." }
+
+        val id = insertIntoUnusedKey(
+                jobs,
+                job,
+                generate = UUID::randomUUID,
+                maxIdle = job.maxBatchDuration,
+                maxIdleUnit = job.maxBatchDurationTimeUnit
+        )
+
+        //We need to let the job know it's own id.
+        jobs.executeOnKey(id) {
+            val v = it.value
+            if (v != null) {
+                v.initId(id)
+                it.setValue(v)
+                if( logger.isDebugEnabled ) logger.debug("Set task id $id")
+            }
+            return@executeOnKey null
+        }
+
+        job.initId(id) //This is duplicate of the job in the map being submitted.
+
+        //Submit to durable executor, grab the task id, and update the job with its task id so that task can retrieve.
+        val f = durableExecutor.submitToKeyOwner(job, id)
+        val taskId = f.taskId
+
+        jobs.executeOnKey(id) {
+            val v = it.value
+            if (v != null) {
+                v.initTaskId(taskId)
+                it.setValue(v)
+                if( logger.isDebugEnabled ) logger.debug("Set task id $taskId for task $id")
```

Comment:
just curious -- is this more performant than just `logger.debug`ging and letting the logger skip over it if debug is not enabled?

Reply:
/should we start adopting this pattern everywhere?

Reply:
It is for kotlin templates, but not for log4j templates. Kotlin templates always compute the string to invoke debug, but if instead of $id we instead {} and passed in id as an argument it wouldn't need if statement. I think either is fine, but think kotlin templates have a slight edge in readability.

Reply:
makes sense! Yeah I prefer the kotlin templating too